collection-loader.js

```diff
@@ -1,42 +1,104 @@
+const get = require('lodash/get');
+const flatMap = require('lodash/flatMap');
 
-function loadCollectionItems(client, collections, {storyFields, storyLimits}) {
-  const bulkRequestBody = collections.reduce(
-    (acc, collection) => Object.assign(acc, {
+function loadCollectionItems(
+  client,
+  collections,
+  {storyFields, storyLimits, defaultNestedLimit}
+) {
+  const bulkRequestBody = collections.reduce((acc, collection) => {
+    let limit = storyLimits[get(collection, ['associated-metadata', 'layout'])];
+
+    if (!limit && get(collection, ['childCollectionLimit'])) {
+      limit = get(collection, ['childCollectionLimit']);
+    }
+
+    if (!limit && defaultNestedLimit) {
+      limit = defaultNestedLimit;
+    }
+
+    return Object.assign(acc, {
       [collection.slug]: {
-        _type: "collection",
+        _type: 'collection',
         slug: collection.slug,
-        "story-fields": storyFields,
-        limit: storyLimits[get(collection, ["associated-metadata", "layout"])]
-      }
-    }), {});
+        'story-fields': storyFields,
+        limit: limit,
+      },
+    });
+  }, {});
 
-  return client.getInBulk({requests: bulkRequestBody}).then(response => response.results);
+  return client
+    .getInBulk({requests: bulkRequestBody})
+    .then((response) => response.results);
 }
 
 // Ugly. This function updates all the items in place.
 // However, this is way more readable than a pure version
-function updateItemsInPlace(client, depth, items, {storyFields, storyLimits}) {
-  const collections = items.filter(item => item && item.type == "collection");
-
-  if(depth == 0 || collections.length == 0)
-    return Promise.resolve();
-
-  return loadCollectionItems(client, collections, {storyFields, storyLimits})
-    .then(collectionSlugToCollection => {
-      collections.forEach(collection => {
-        collection.summary = get(collectionSlugToCollection, [collection.slug, "summary"], '')
-        collection.items = get(collectionSlugToCollection, [collection.slug, "items"])
-      });
-      return updateItemsInPlace(client, depth - 1, flatMap(collections, collection => collection.items), {storyFields, storyLimits})
-    })
+function updateItemsInPlace(
+  client,
+  depth,
+  items,
+  {storyFields, storyLimits, defaultNestedLimit, nestedCollectionLimit}
+) {
+  const collections = items.filter((item) => item && item.type == 'collection');
+
+  if (depth == 0 || collections.length == 0) return Promise.resolve();
+
+  return loadCollectionItems(client, collections, {
+    storyFields,
+    storyLimits,
+    defaultNestedLimit,
+  }).then((collectionSlugToCollection) => {
+    collections.forEach((collection) => {
+      collection.summary = get(
+        collectionSlugToCollection,
+        [collection.slug, 'summary'],
+        ''
+      );
+      collection.items = get(collectionSlugToCollection, [
+        collection.slug,
+        'items',
+      ]);
+    });
+
+    collections.forEach((collection) =>
```

Comment:
There would definitely be an efficiency issue

Reply:
removed filter